src/lib/device/journal_vdev.cpp

```diff
@@ -34,75 +34,211 @@
+            m_dmgr.remove_chunk(chunk);
+        }
+    }
 
-        // Update the overhead to total write size;
-        m_write_sz_in_total.fetch_add(chunk->size() - offset_in_chunk, std::memory_order_relaxed);
+    // Start the chunk pool.
+    m_chunk_pool->start();
+}
 
-        // If across chunk boundary, update the chunk super-block of the chunk size
-        chunk->update_end_of_chunk(offset_in_chunk);
+void JournalVirtualDev::update_chunk_private(shared< Chunk >& chunk, JournalChunkPrivate* private_data) {
+    sisl::blob blob;
+    blob.bytes = r_cast< uint8_t* >(private_data);
+    blob.size = sizeof(JournalChunkPrivate);
+    chunk->set_user_private(blob);
+}
 
-#ifdef _PRERELEASE
-        iomgr_flip::test_and_abort("abort_after_update_eof_cur_chunk");
-#endif
-        // get next chunk handle
-        auto next_chunk = get_next_chunk(chunk);
-        if (next_chunk != chunk) {
-            // Since we are re-using a new chunk, update this chunk's end as its original size;
-            next_chunk->update_end_of_chunk(chunk->size());
+uint64_t JournalVirtualDev::get_end_of_chunk(shared< Chunk >& chunk) const {
+    auto* private_data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
+    return private_data->end_of_chunk_size;
+}
+
+shared< JournalVirtualDev::Descriptor > JournalVirtualDev::open(logdev_id_t logdev_id) {
+    auto it = m_journal_descriptors.find(logdev_id);
+    if (it == m_journal_descriptors.end()) {
+        auto journal_desc = std::make_shared< JournalVirtualDev::Descriptor >(*this, logdev_id);
+        m_journal_descriptors.emplace(logdev_id, journal_desc);
+        return journal_desc;
+    }
+
+    HS_LOG(TRACE, device, "Opened log device descriptor {}", logdev_id);
+    return it->second;
+}
+
+void JournalVirtualDev::Descriptor::append_chunk() {
+    // Get a new chunk from the pool.
+    auto new_chunk = m_vdev.m_chunk_pool->dequeue();
+
+    // Increase the right window and total size.
+    m_total_size += new_chunk->size();
+    m_end_offset += new_chunk->size();
+
+    if (!m_journal_chunks.empty()) {
+        // If there are already chunks in the m_journal_chunks list, append this new chunk to the end of the list. Write
+        // the next_chunk of the last chunk in the list to point to this new chunk. If already there are no chunks make
+        // the new chunk as the head.
+        auto last_chunk = m_journal_chunks.back();
+        auto* last_chunk_private = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(last_chunk->user_private()));
+
+        // Set the next chunk with the newly created chunk id.
+        last_chunk_private->next_chunk = new_chunk->chunk_id();
+
+        // Append the new chunk
+        m_journal_chunks.push_back(new_chunk);
+        auto chunk_size = m_vdev.info().chunk_size;
+        auto offset_in_chunk = (tail_offset() % chunk_size);
```

Comment:
suggest to call offset_to_chunk to not assume chunk is always same size.

Reply:
As discussed all chunks in a vdev have same size.